src/core/friendly_errors/param_validator.js

```diff
@@ -313,6 +318,114 @@ function validateParams(p5, fn) {
+    // (after scoring the schema closeness in `findClosestSchema`). Here, we
+    // always print the first error so that user can work through the errors
+    // one by one.
+    let currentError = zodErrorObj.errors[0];
+
+    // Helper function to build a type mismatch message.
+    const buildTypeMismatchMessage = (actualType, expectedTypeStr, position) => {
+      const positionStr = position ? `at the ${ordinals[position]} parameter` : '';
+      const actualTypeStr = actualType ? `, but received ${actualType}` : '';
+      return `Expected ${expectedTypeStr} ${positionStr}${actualTypeStr}.`;
+    }
+
+    // Union errors occur when a parameter can be of multiple types but is not
+    // of any of them. In this case, aggregate all possible types and print
+    // a friendly error message that indicates what the expected types are at
+    // which position (position is not 0-indexed, for accessibility reasons).
+    const processUnionError = (error) => {
+      const expectedTypes = new Set();
+      let actualType;
+
+      error.unionErrors.forEach(err => {
+        const issue = err.issues[0];
+        if (issue) {
+          if (!actualType) {
+            actualType = issue.received;
+          }
+
+          if (issue.code === 'invalid_type') {
+            expectedTypes.add(issue.expected);
+          }
+          // The case for constants. Since we don't want to print out the actual
+          // constant values in the error message, the error message will
+          // direct users to the documentation.
+          else if (issue.code === 'invalid_literal') {
+            expectedTypes.add("constant (please refer to documentation for allowed values)");
+          } else if (issue.code === 'custom') {
+            const match = issue.message.match(/Input not instance of (\w+)/);
+            if (match) expectedTypes.add(match[1]);
+          }
+        }
+      });
+
+      if (expectedTypes.size > 0) {
+        const expectedTypesStr = Array.from(expectedTypes).join(' or ');
+        const position = error.path.join('.');
+
+        message = buildTypeMismatchMessage(actualType, expectedTypesStr, position);
+      }
+
+      return message;
+    }
+
+    switch (currentError.code) {
+      case 'invalid_union': {
+        processUnionError(currentError);
+        break;
+      }
+      case 'too_small': {
+        const minArgs = currentError.minimum;
+        message = `Expected at least ${minArgs} argument${minArgs > 1 ? 's' : ''}, but received fewer. Please add more arguments!`;
+        break;
+      }
+      case 'invalid_type': {
+        message = buildTypeMismatchMessage(currentError.received, currentError.expected, currentError.path.join('.'));
+        break;
+      }
+      case 'too_big': {
+        const maxArgs = currentError.maximum;
+        message = `Expected at most ${maxArgs} argument${maxArgs > 1 ? 's' : ''}, but received more. Please delete some arguments!`;
```

Comment:
The messages probably should have the name of the function being called in there to help with identifying source of error, if it is not already part of the message. 

The suggested fix likely need to be more of a suggestive tone as the suggested fix may not be the actual problem. ie. instead of "Please add/delete arguments" go with something more like "Did you forget to add an argument?"/"Removing some arguments may help", or the suggestion can be omitted as well in line with the current FES's behavior.

Reply:
Sounds good!